pageserver/src/tenant_mgr.rs

```diff
@@ -379,20 +374,14 @@ pub fn get_local_timeline_with_load(
     tenant_id: ZTenantId,
     timeline_id: ZTimelineId,
 ) -> anyhow::Result<Arc<Timeline>> {
-    let mut m = tenants_state::write_tenants();
-    let tenant = m
-        .get_mut(&tenant_id)
-        .with_context(|| format!("Tenant {tenant_id} not found"))?;
-
-    if let Some(page_tline) = tenant.local_timelines.get(&timeline_id) {
-        Ok(Arc::clone(page_tline))
-    } else {
-        let page_tline = load_local_timeline(&tenant.repo, timeline_id)
-            .with_context(|| format!("Failed to load local timeline for tenant {tenant_id}"))?;
-        tenant
-            .local_timelines
-            .insert(timeline_id, Arc::clone(&page_tline));
-        Ok(page_tline)
+    let repository = get_repository_for_tenant(tenant_id)?;
+    match repository.get_timeline(timeline_id) {
+        Some(RepositoryTimeline::Loaded(loaded_timeline)) => {
+            loaded_timeline.init_logical_size()?;
```

Comment:
Note this change: looks like this is the most crucial thing in the current timeline management in pageserver, since apparently straightens out the timeline's state.

Here's a set of requirements every timeline has to fulfil to become properly loaded:

* timeline's `init_logical_size` has to be called before any public interaction outside of `Repository` is made with the timeline
* we cannot call `init_logical_size` right away in `get_timeline_load_internal` or `load_local_timeline`, we'll get an error
* a timeline has to start walreceiver when its layers are loaded into memory

The old code, using `local_timelines` map, implicitly calling `init_logical_size` and enabling walreceiver only once per timeline, using timeline's presence in `local_timelines` as "had not been initialized before public access" trait.

Now, when we've removed all the maps and timelines' separation and about to remove timeline load state and overhaul the rest of the logic, what would be the best way to handle this?

I could have added some `walreceiver_started: AtomicBool` in `Timeline` and check it here, but looks like we should rather change the logic so that unitialized timeline does not get out `Repository`, wdyt @hlinnaka ?

If constant `init_logical_size` calls here are ok, we can continue with refactorings later and merge this for now.

Reply:
> I could have added some walreceiver_started: AtomicBool in Timeline and check it here, but looks like we should rather change the logic so that unitialized timeline does not get out Repository, wdyt @hlinnaka ?

Hmm. There are multiple stages to initializing a timeline:
1. download index file and all layer files to local disk, and build the corresponding in-memory structs
2. initialize logical size
3. start walreceiver

Step 3 isn't actually required for queries. GetPage requests can be answered without a WAL receiver. It's only required for receiving new WAL. 

I'm not sure if step 2 is needed for GetPage requests either. We might have some implicit assumptions on that currently, but in principle reconstructing a page doesn't require knowing the logical size.

I think the best long term solution is to calculate the logical size lazily. With that, we could start serving GetPage requests and also launch the WAL receiver immediately after step 1. The "has WAL receiver been launched?" and "has logical size been calculated?" states could be separate independent flags.

> If constant init_logical_size calls here are ok, we can continue with refactorings later and merge this for now.

`init_logical_size` is pretty expensive, you don't want to call it more often than necessary. Perhaps add a flag for "has logical size been initialized?", and short-circuit `init_logical_size` so that if the size has already been initialized, do nothing.

Reply:
I would also imagine, we could refactor the code more to ensure those init calls do happen explicitly and no one is able to miss them before working with the struct, but I think this better happen after the overhaul PR.

Meanwhile as a stub, I'll add the "has logical size been initialized" flag or similar.

Reply:
Ok. I'm not 100% it's free from race conditions, but that's probably good enough for now